bug: PhpdocSeparationFixer - Make groups handling more flexible

julienfalque opened the pull request:
Fixes #6635.

This is a proposal to change the way the option works a bit.

Before, two tags were grouped together if they were both explicity listed in the same group and separated otherwise.

Now, they are still grouped if they are both in the same group, but they are separated *only* if one is in a group and the other is in another group or not in a group at all. If both are not in any group they are left untouched, meaning they remain grouped or separated.

So if we want to separate a single tag from others, we now have to put it explicity in a group too. This is kind of a BC break though, but I think it's reasonable given the fact that the current behavior is a bit unexpected.

This PR also adds support for more complex tags such as Doctrine annotations; and deprecates `TagComparator` which implies a minor release if merged.

julienfalque commented on tests/AutoReview/ProjectCodeTest.php:
```diff
@@ -526,34 +525,6 @@ public function testAllCodeContainSingleClassy(string $className): void
-    /**
-     * @dataProvider provideSrcClassCases
-     */
-    public function testThereIsNoTriggerErrorUsedDirectly(string $className): void
```
The added deprecation in this PR is not intended for end-users but for library authors. I don't know how to keep this test and tell it which `trigger_error()` calls to allow reliably.

julienfalque commented on src/DocBlock/TagComparator.php:
```diff
@@ -42,6 +44,8 @@ final class TagComparator
      */
     public static function shouldBeTogether(Tag $first, Tag $second, array $groups = self::DEFAULT_GROUPS): bool
     {
+        @trigger_error('Method '.__METHOD__.' is deprecated and will be removed in version 4.0.');
```
This should use `E_DEPRECATED`. Also test missing.

VincentLanglet commented:
> So if we want to separate a single tag from others, we now have to put it explicity in a group too. This is kind of a BC break though

This would be an important and annoying BC-break for static-analysis user which phpdoc like `phpstan-param`, `psalm-param`, `phpstan-return`, `phpstan-template` ...

Adding all the existing annotation to the config is way too complicated.

> but I think it's reasonable given the fact that the current behavior is a bit unexpected.

The description of the rule is 
```
Annotations in PHPDoc should be grouped together so that annotations of the same type immediately follow each other. Annotations of a different type are separated by a single blank line.
```

So it's expected. "Annotations of a different type are separated by a single blank line." is meaning that any different annotation are separated by one line, there was nothing explaining it was only the annotation from the groups.

IMHO, there are multiple things in the link issue
```
@Assert\NotBlank
@Assert\Email

@OA\Property(type="string")

@Method('GET')
```
First, it would be interesting to understand why the spaces are remove between those tags ; which are the opposite behavior of what is expected. I would expect instead instead
```
@Assert\NotBlank

@Assert\Email

@OA\Property(type="string")

@Method('GET')
```
Then, since people are generally grouping the Assert\* together, we should provide an easy way to do this (adding all the Assert\* to the config is too complicated). I see two solutions
- Considering that the tag content is only the part before the first `\`, so we would have this behavior by default
- Adding support for regex in the config, so it would be possible to add `['Assert\*']` to group all the asserts.

julienfalque commented:
@VincentLanglet You're right, I reworked the changes to add a `skip_unlisted_annotations` option. The behavior I described above only applies when the option is enabled. I still think this should be the default behavior so I propose to set the default value of this option to `true` in next major release.

> First, it would be interesting to understand why the spaces are remove between those tags ; which are the opposite behavior of what is expected

This was because `\` was not allowed in the tag names. And unreadble tags were always considered to be in the same group. Now `\` is allowed and in case a tag still cannot be read, it should be skipped.

Finally, I added support for `*` wildcard in tag names.

All those changes should give enough flexibility to tweak the rule to one's needs but it will likely require updating the configuration when updating the tool to avoid adding lines between tags such as `@ORM\…` or `@Assert\…`.

VincentLanglet commented:
> @VincentLanglet You're right, I reworked the changes to add a `skip_unlisted_annotations` option. The behavior I described above only applies when the option is enabled. I still think this should be the default behavior so I propose to set the default value of this option to `true` in next major release.
> 
> > First, it would be interesting to understand why the spaces are remove between those tags ; which are the opposite behavior of what is expected
> 
> This was because `\` was not allowed in the tag names. And unreadble tags were always considered to be in the same group. Now `\` is allowed and in case a tag still cannot be read, it should be skipped.
> 
> Finally, I added support for `*` wildcard in tag names.
> 
> All those changes should give enough flexibility to tweak the rule to one's needs but it will likely require updating the configuration when updating the tool to avoid adding lines between tags such as `@ORM\…` or `@Assert\…`.

Awesome changes. This seems to be a perfect solution.

keradus commented:
Big thanks!
